### PR TITLE
wasmtime: In-process sampling profiler

### PR DESCRIPTION
**Pull request by jameysharp**
This isn't ready for detailed review but the basic functionality seems to work so I'm putting it up in case anyone wants to play with it.

Unlike the existing profiling options, this works on all platforms and does not rely on any external profiling tools like perf or VTune. On the other hand, it can only profile time spent in the WebAssembly guest, not in Wasmtime itself or other host code. Also it can't measure time as precisely as platform-native tools can.

### CODE REVIEW

**Comment by jameysharp** [comment]
This PR now produces output that works as expected with https://profiler.firefox.com/, and I don't think it's fundamentally missing anything.

Before merging, I'd like to move the `GuestProfiler` struct somewhere in the `wasmtime` crate, instead of leaving it in `wasmtime-cli` where only `wasmtime run` can use it.

I don't think any of the implementation will change though, just move elsewhere, so I'm marking this as ready for review.

File: src/commands/run.rs
Comment (alexcrichton) [inline src/commands/run.rs:474]: I think that the error handling here may want to be a bit more nuanced perhaps. For example if `invoke_func` failed and this finalization also fails, the original failure will be masked by the writing of the profile. 

This concern of mine is additionally shared with the writing of a core dump after a trap, which IIRC will print out an error but won't return the core dump writing error further up the stack. Could the logic there and here be unified perhaps into a sort of "perform a fallible operation post-wasm-execution"?

File: src/commands/run.rs
Comment (alexcrichton) [inline src/commands/run.rs:392]: Could this be moved into a CLI argument as well?

File: src/commands/run.rs
Comment (alexcrichton) [inline src/commands/run.rs:643]: The interning and various allocations here make me wonder: would it be better to defer this until after execution? Should this instead buffer up a `Vec<WasmBacktrace>` which is then all serialized out at the end of execution? 

Or perhaps amortizing the cost by collecting N backtraces and then flushing them all at once perhaps.

File: src/commands/run.rs
Comment (alexcrichton) [inline src/commands/run.rs:639]: Are there any flags or configuration we could perhaps pass to indicate which module a frame comes from? Or perhaps which instance a frame comes from? We don't currently have a means of learning that from a wasm frame but it's something we could theoretically add.

File: src/commands/run.rs
Comment (jameysharp) [reply src/commands/run.rs:657]: I want to move the `GuestProfiler` type into the `wasmtime` crate primarily because I want it to be usable by embedders, but an additional reason is because then it's reasonable for it to use `wasmtime-runtime`'s low-level stack-walking API directly. I think there's a lot of work that can be avoided that way.

I don't think batching up `WasmBacktrace` objects is going to help much, because just collecting one of those involves looking up which function each PC comes from, along with information like the function name. At that point I'd guess three hash-table lookups per function for interning aren't going to make much difference.

I'd like to merge this with the current approach, of eagerly consuming the high-level `WasmBacktrace`, as-is for now. I don't think the public API of `GuestProfiler` will need to change to support more efficient collection of backtraces, so I think we can iterate on performance improvements afterward without impacting any users.

I definitely want to improve this over time though. I think eventually we're going to want API changes in `fxprof-processed-profile` for the highest performance sampling we can do in-process.

File: crates/wasmtime/src/profiling.rs
Comment (alexcrichton) [inline crates/wasmtime/src/profiling.rs:27]: I think this'd be a good place to have a doc-block example if you're up for it about how to configure a `Store` and such.

File: src/commands/run.rs
Comment (alexcrichton) [inline src/commands/run.rs:402]: API question: instead of requiring an `Arc` and `Mutex` here, which technically aren't necessary since there aren't any threads, could this instead be stored in the `T` of `Store<T>`? As-is the `GuestProfiler::sample` method would not work since it would be stored in `T` but additionally takes a store as a parameter, but I think that could be fixed by taking a`&WasmBacktrace` as an argument since that's all the store is used for internally.

**Review by alexcrichton — COMMENTED** [review]
I'm ok landing this personally, but I think we should be clear in the documentation of the type that it's not really at a state of implementation quality to be a high-fidelity general purpose guest profiler. For example some drawbacks I can think of are:

* Requires epochs as the driver for stack collection, meaning the granularity of samples is function entries and loop headers. Additionally this means that you're not necessarily measuring the "raw" performance since it's code-based instrumentation.
* Stacks are always collected from wasm meaning that any time spent in the host is not accounted for. This means that samples can be lost while in the host and additionally whether or not the host blocked or did compute I think is unclear. (although perhaps this is fixable)

I suspect that a large number of use cases are totally ok with these drawbacks, but at the same time I think it's important to be clear that we're not looking to get into the business of a "real" profiler at this time with off-thread stack sampling, host integration, etc.

Additionally would you also be up for updating the documentation on `docs/*.md`? Currently we have `examples-profiling.md` which delegates to `perf` and `vtune` and I think it'd be good to add an entry for this strategy of profiling. Ideally there could be at least a small paragraph of words explaining why one might choose one profiling strategy over another (e.g. `perf` captures everything but is Linux-specific and not always easy to interpret, while this is cross-platform but not as "high fidelity" as `perf`)

File: src/commands/run.rs
Comment (fitzgen) [inline src/commands/run.rs:439]: Doesn't matter a ton in this case, but one pattern I've used to make APIs like this slightly harder to get wrong is to return an `impl Drop`:

```rust
struct FinishEpochHandler<F: FnOnce()>(Option<F>);
impl<F: FnOnce()> Drop for FinishEpochHandler<F> {
    fn drop(&mut self) {
        if let Some(f) = self.0.take() {
            f();
        }
    }
}
return FinishEpochHandler(Some(move || {
    // what you have now...
}));
```

This way, callers don't need to actually remember to call the resulting function, at most they just need to add a scope.

But I guess if you do Alex's suggestion of moving the profiler into `Store`, then this result function would require a store argument, and the RAII/`impl Drop` approach won't really work anymore...

**Review by fitzgen — COMMENTED** [review]
Really excited for this!

When profiling is enabled in the CLI, it probably makes sense to write some output after running the Wasm that repeats where the profile file was written to and how to upload it to `https://profiler.firefox.com/`

File: src/commands/run.rs
Comment (jameysharp) [reply src/commands/run.rs:402]: "Wait, I could just keep it in the Store," I thought, as I fell asleep last night; then wake up to find this comment. :laughing: 

I don't want to make `WasmBacktrace` part of the public API here because I want to re-implement this in terms of the lower-level `wasmtime-runtime` stack-walking API. That doesn't directly need access to the store at all, but I do need some information from the store at some point to figure out how native PCs map to individual modules and to functions within those modules.

It might work though to pass in the store when constructing a `GuestProfiler`, collect just the necessary information about all the modules and symbols from it, and then not need it in `sample`.

Otherwise, the only alternative I've thought of is to define `fn sample(store: impl AsContextMut, get_self: impl FnOnce(&mut Store) -> &mut Self)`. Not the nicest interface but I think it should work, right?

File: src/commands/run.rs
Comment (jameysharp) [reply src/commands/run.rs:175]: I've only been using the "local file" interface, so I didn't realize there was a way to upload a profile and share a link. I guess the way to do that is documented at https://github.com/firefox-devtools/profiler/blob/main/docs-developer/loading-in-profiles.md.

I don't see any HTTP client usage anywhere in Wasmtime except in wasi-http right now. So that seems like a relatively large change that I'd prefer to think about separately from the basic profiling implementation. It would definitely be nice for usability, though.

File: crates/wasmtime/src/profiling.rs
Comment (jameysharp) [reply crates/wasmtime/src/profiling.rs:96]: If you call `finish` again it'll just update the thread end-time and write out the whole profile again, including any new samples. So it's probably not useful to add more samples after calling `finish`, but it should be harmless.

That said, I only did it this way because there's still an extra reference to the `GuestProfiler` in the epoch callback after the instance finishes, and I didn't want to have to clone the profile. But if I can make Alex's suggestion work, of keeping the `GuestProfiler` in the store instead of in an `Arc`, then that won't be an issue and I could make the change you're suggesting too.

File: src/commands/run.rs
Comment (alexcrichton) [reply src/commands/run.rs:175]: Ah ok yeah if it's not easy to reuse whatever is in wasi-http I'm happy to leave this as-is

File: src/commands/run.rs
Comment (alexcrichton) [reply src/commands/run.rs:402]: How about:

```rust
impl<T> Store<T> {
    pub fn sample(&mut self, get_profiler: impl FnOnce(&mut T) -> &mut GuestProfiler) { ... }
}
```

with similar impls for `StoreContextMut`?

which is more-or-less the same thing, but just as a method instead of a function on `GuestProfiler` now that I reread your comment

**Comment by jameysharp** [comment]
I got rid of the need for `Store` in `sample()`. Then I discovered that I can't put a `GuestProfiler` directly in the `Host` structure because some of the WASI extensions require that `Host` must be cloneable.

As far as I can tell I end up needing an `Arc` and `Mutex` anyway. I would love for somebody to tell me how to avoid that.

It was still worth doing this because now any particular PC is only looked up once per profile and cached by `fxprof_processed_profile`. Different PCs within the same function need to be looked up separately but since we only collect stack traces at epoch interruption points there aren't so many distinct PCs. This should significantly reduce profiling overhead but I haven't measured how much impact it actually has.

I struggled with this a bunch because I was trying to make it automatically use all modules which have been registered within the current `Store`'s `ModuleRegistry`. But for WASI commands the module isn't registered in the store until you actually try to invoke it, and I tried all sorts of ways to hook into that registration before realizing that the caller knows which modules they care about and I can just make it their problem to tell me.

Then I decided that it's a feature to be able to exclude modules from profiles: if you have some proprietary module that you're linking with your customer's modules, you may not want to reveal anything about the implementation of your module when providing them with a profile.

**Comment by alexcrichton** [comment]
> As far as I can tell I end up needing an Arc and Mutex anyway. 

Ah this is due to the integration of `wasi-threads` where that's modeled by cloning the store's context to duplicate it on another thread. Otherwise though the cloning is not required. You can take a similar route as the `WasiNnCtx` integration, for example, which additionally does not implement `Clone` so what it does is it's bundled into an `Arc` where whenever it's required it calls `Arc::get_mut`. That fails if there's more than one clone, or more than one thread, and the error message says "this doesn't support threads yet". I think that profiling will fall into the same bucket?

In the long-term future I could imagine that there's a standalone profiler struct where you acquire thread-specific profilers from that structure. The `Host` could store the standalone profiler (which internally probably has an arc and a mutex) and additionally thread-specific profilers. The `Clone for Host` impl would then use the standalone profiler to get a new thread-specific profiler, allowing each thread to have mutable access to its profiler but no others. Something like that. No need to do here of course, mostly just modeling how this may look in the future with support for threads.

> I can just make it their problem to tell me.

This seems like a neat idea to me! I like the idea of precomputing symbol tables. I do think this has the downside of that it has more setup required, but that's also perhaps not a bad thing and it's not exactly onerous at this time, so it seems fine by me.

> I got rid of the need for Store in sample().

Personally I'm a bit wary of this because to me I wouldn't want to commit long-term to being able to capture a stack trace without any input. I'm not sure how we'll ever move away from those thread locals but I sure would love to one day if someone gets a good idea. In such a situation there's no way to start the backtrace without a contextual store being passed  in. All that being said though it's not like we need to get the API 100% prefect to begin with, that's why we have major version bumps after all. I mostly wanted to put this out there, but I additionally think it's fine to land as-is with your current strategy. If and when we remove TLS (fingers crossed) we can figure out how to change this API at this point. I realize that passing the store into something stored within the store is not easy, so let's stick with what you've got.

**Comment by jameysharp** [comment]
Thanks for the tips, Alex: that worked!

Okay, I've addressed most of the issues that you all have raised. What I haven't done is added documentation to the Wasmtime book, or API usage examples to the doc-comments. I'd kind of like to merge the implementation and then come back to that, if that's okay. I hope the doc-comments I've written so far will be a good start for the book, and the usage in `run.rs` will be a good start for the usage examples…

There are also still several to-do items, like measuring CPU time and noting wasm entry/leave events, which I also want to defer to future work.

**Review by alexcrichton — APPROVED** [review]
Sounds like a good plan to me 👍

**Pull request merged**
